### DISCUSSION TITLE
Tips and procedure to add new APIs

### DISCUSSION THREAD

**Discussion by shockwaveHe** [microsoft/AirSim#3606] [Support Q&A] [answered]
Hi! My project on AirSim requires me to add some new APIs for our own need. I have tried to add new APIs by myself, but still some bugs exist. I checked the guide to add new APIs proposed on AirSim offcial website, but it seems that it is out of date. 
Can you upload the guide to add new APIs, especially world APIs?

**Comment by zimmy87** [comment, marked as answer]
Hi @shockwaveHe, #3102 is an example of adding a new world API. Also, is the guide you're looking at for adding new APIs the one located [here](https://microsoft.github.io/AirSim/apis/#adding-new-apis-to-airsim) or are you looking elsewhere?

**Reply by shockwaveHe** [reply]
Hi @zimmy87, thanks for your help! I will try to follow a similar procedure to implement the new API. 
The guide I am looking at is exactly what you mentioned.
Besides, we should run build.cmd after adding the new API, right?

On more question, is there a way to get the simulation time inside AirSim or UE?

**Reply by rajat2004** [reply]
`ClockFactory::get()->nowNanos()` should work for getting the time in nanoseconds, there mostly will be examples of how to use that in the files itself

**Reply by shockwaveHe** [reply]
Thanks a lot! @rajat2004

**Reply by rajat2004** [reply]
Forgot to update here, there's a PR #3611 which updates the code structure, and adds a new page on adding APIs with some example links

**Reply by shockwaveHe** [reply]
@rajat2004 Cool! Actually I figured out the how to add APIs before the reply, but still thanks your effort!
